peps/pep-0733.rst

```diff
@@ -0,0 +1,655 @@
+  semantics, virtual multiple inheritance, 1-based indexing, super-long
+  inheritance chains, goroutines, channels, etc.).
+
+Strengths of the C API
+======================
+
+While the bulk of this document is devoted to problems with the
+C API that we would like to see fixed in any new design, it is
+also important to point out the strengths of the C API, and to
+make sure that they are preserved.
+
+As mentioned in the introduction, the C API enabled the
+development and growth of the Python ecosystem over the last
+three decades, while evolving to support use cases that it was
+not originally designed for. This track record in itself is
+indication of how effective and valuable it has been.
+
+A number of specific strengths were mentioned in the
+capi-workgroup discussions. Heap types were identified
+as much safer and easier to use than static types
+[`Issue 4 <https://github.com/capi-workgroup/problems/issues/4#issuecomment-1542324451>`__].
+
+API functions that take a C string literal for lookups based
+on a Python string are very convenient
+[`Issue 30 <https://github.com/capi-workgroup/problems/issues/30#issuecomment-1550098113>`__].
+
+The Limited API and stable ABI hide implementation details and
+make it easier to evolve Python
+[`Issue 30 <https://github.com/capi-workgroup/problems/issues/30#issuecomment-1560083258>`__].
+
+C API problems
+==============
+
+The remainder of this document summarizes and categorizes the problems that were reported on
+the `capi-workgroup <https://github.com/capi-workgroup/problems/issues/>`__ repository.
+The issues are grouped into several categories.
+
+
+API Evolution and Maintenance
+-----------------------------
+
+The difficulty of making changes in the C API is central to this report. It is
+implicit in many of the issues we discuss here, particularly when we need to
+decide whether an incremental bugfix can resolve the issue, or whether it can
+only be addressed as part of an API redesign
+[`Issue 44 <https://github.com/capi-workgroup/problems/issues/44>`__]. The
+benefit of each incremental change is often viewed as too small to justify the
+disruption. Over time, this implies that every mistake we make in an API's
+design or implementation remains with us indefinitely.
+
+We can take two views on this issue. One is that this is a problem and the
+solution needs to be baked into any new C API we design, in the form of a
+process for incremental API evolution. The other possible approach is that
+this is not a problem to be solved, but rather a feature of any API. In this
+view, API evolution should not be incremental, but rather through large
+redesigns, each of which learns from the mistakes of the past and is not
+shackled by backwards compatibility requirements. A compromise approach
+is somewhere between these two extremes, fixing issues which are easy
+or important enough to tackle incrementally, and leaving others alone.
```

Comment:
This paragraph presupposes a trade-off which is not necessarily necessary.

With ABI versioning, it is possible for a Python implementation to concurrently support two very different ABIs and so have the best of both worlds. HPy currently implements such a solution.

Maybe one does not like a particular solution, but this isn't a simple zero sum slider.

I don't have a particular edit to suggest, but would we happy if a sentence was added. The paragraph after this one does a great job of describing the crux of the matter -- we needed an agreed upon plan for API evolution.

Reply:
> With ABI versioning, it is possible for a Python implementation to concurrently support two very different ABIs and so have the best of both worlds. 

This is what you do in both the evolutionary and the revolutionary options. There is no option where you evolve the API/ABI without versioning. The proponents of stability argue that adding new things is fine, but nothing should ever be removed and nobody should ever be forced off the old API. The evolutionary approach is that you do eventually deprecate and remove old version of the API, so that you don't need to maintain them forever.

Reply:
I changed the text to clarify this.